src/xmerl_c14n.erl

```diff
@@ -96,21 +96,23 @@ needed_ns(#xmlElement{nsinfo = NsInfo, attributes = Attrs}, InclNs) ->
         %[] -> [K || {K,V} <- E#xmlElement.namespace#xmlNamespace.nodes];
         %_ -> NeededNs1
     %end,
-    lists:foldl(fun(Attr, Needed) ->
-        case Attr#xmlAttribute.nsinfo of
-            {"xmlns", Prefix} ->
-                case lists:member(Prefix, InclNs) of
-                    true -> [Prefix | Needed];
-                    _ -> Needed
-                end;
-            {Ns, _Name} ->
-                case lists:member(Ns, Needed) of
-                    true -> Needed;
-                    _ -> [Ns | Needed]
-                end;
-            _ -> Needed
-        end
-    end, NeededNs2, Attrs).
+    All_NS =
+        lists:foldl(fun(Attr, Needed) ->
+            case Attr#xmlAttribute.nsinfo of
+                {"xmlns", Prefix} ->
+                    case lists:member(Prefix, InclNs) of
+                        true -> [Prefix | Needed];
+                        _ -> Needed
+                    end;
+                {Ns, _Name} ->
+                    case lists:member(Ns, Needed) of
+                        true -> Needed;
+                        _ -> [Ns | Needed]
+                    end;
+                _ -> Needed
+            end
+        end, NeededNs2, Attrs),
+    sets:to_list(sets:from_list(All_NS)).
```

Comment:
@zvoykish 
in the case match for the second clause - its checking lists:member(Ns, Needed)
why can' we do the same for the first clause instead?

Reply:
@nitzanharel 
I admit I haven't looked into optimizing the logic itself, and wanted to commit as minor changes as possible.
Looking into it - I don't think we can since the 1st case is looking and adding the 2nd tuple element and the 2nd case is using the 1st tuple element.

Reply:
@zvoykish - maybe I wasn't clear so apologies.
I was thinking of changing:
case lists:member(Prefix, InclNs) of
                        true -> [Prefix | Needed];
                        _ -> Needed
                    end;

to:
case lists:member(Prefix, InclNs) of
                        true -> 
           case lists:member(Prefix, Needed) of
               true -> Needed;
               false -> [Prefix | Needed]
           end;
                        _ -> Needed
                    end;